mabwiser/treebandit.py

```diff
@@ -0,0 +1,163 @@
+            # consecutive/follow numerical order, but are always representative of leaf nodes.
+            leaf_indices = self.arm_to_tree[arm].apply(arm_contexts)
+
+            # Use set() to create a list of unique indices
+            # These indices represent the leaves reached via the given contexts
+            unique_leaf_indices = set(leaf_indices)
+
+            for index in unique_leaf_indices:
+                # Get rewards list for each leaf
+                self.arm_to_rewards[arm][index] = arm_rewards[leaf_indices == index]
+
+    def _predict_contexts(self, contexts: np.ndarray, is_predict: bool,
+                          seeds: Optional[np.ndarray] = None, start_index: Optional[int] = None) -> List:
+
+        # Get local copy of arm_to_tree, arm_to_expectation, arm_to_rewards, and arms
+        # to minimize communication overhead between arms (processes) using shared objects
+        arm_to_tree = deepcopy(self.arm_to_tree)
+        arm_to_rewards = deepcopy(self.arm_to_rewards)
+        arm_to_expectation = deepcopy(self.arm_to_expectation)
+        arms = deepcopy(self.arms)
+
+        # Create an empty list of predictions
+        predictions = [None] * len(contexts)
+        for index, row in enumerate(contexts):
+            # Each row needs a separately seeded rng for reproducibility in parallel
+            rng = np.random.RandomState(seed=seeds[index])
+
+            for arm in arms:
+
+                # Copy the row rng to the deep copied model in arm_to_tree
+                arm_to_tree[arm].rng = rng
+
+                # If there was prior data for this arm, get expectation for arm
+                if arm_to_rewards[arm]:
+                    # Get leaf index for that context
+                    leaf_index = arm_to_tree[arm].apply([row])[0]
+
+                    # Get the rewards list of that leaf
+                    leaf_rewards = arm_to_rewards[arm][leaf_index]
+
+                    # Create leaf lp
+                    leaf_lp = self._get_leaf_lp(arm)
+
+                    # Leaf LP: fit the same arm decision with the leaf rewards
+                    leaf_lp.fit(np.asarray([arm] * len(leaf_rewards)), leaf_rewards)
+
+                    # Leaf LP: predict expectation
+                    arm_to_expectation[arm] = leaf_lp.predict_expectations()[arm]
+
+            if is_predict:
+                predictions[index] = argmax(arm_to_expectation)
+            else:
+                predictions[index] = arm_to_expectation.copy()
+
+        # Return list of predictions
+        return predictions
+
+    def _uptake_new_arm(self, arm: Arm, binarizer: Callable = None, scaler: Callable = None):
+
+        self.lp.add_arm(arm, binarizer)
+        self.arm_to_tree[arm] = DecisionTreeClassifier(**self.tree_parameters)
+        self.arm_to_rewards[arm] = dict()
+
+    def _get_leaf_lp(self, arm: Arm):
+
+        # TODO: maybe we don't need re-create but just change self.lp.arms to [arm, -1]?
```

Comment:
did you get a chance to see, if simply revising the arms of the original MAB.lp object here produces identical results? 

If so, maybe we don't need to allocated a new LP for each region and instead, all leaf regions uses the "same" LP object that comes from MAB.LP.. which would be self.lp here. 

Of course, creating new objects each time seems "safer" but I wonder if the other is semantically equivalent..

Reply:
I checked and it unfortunately doesn't work. Changing the original lp creates problems, so it's probably better to keep _get_leaf_lp().